pkg/component/runtime/command.go

```diff
@@ -392,8 +396,18 @@ func (c *CommandRuntime) handleProc(state *os.ProcessState) bool {
 	switch c.actionState {
 	case actionStart:
 		// should still be running
-		stopMsg := fmt.Sprintf("Failed: pid '%d' exited with code '%d'", state.Pid(), state.ExitCode())
-		c.forceCompState(client.UnitStateFailed, stopMsg)
+		reportFailure := true
+		if c.restartBucket != nil {
+			// report failure only if bucket is full of restart events
+			reportFailure = !c.restartBucket.Allow()
+		}
+
+		if reportFailure {
+			stopMsg := fmt.Sprintf("Failed: pid '%d' exited with code '%d'", state.Pid(), state.ExitCode())
+			c.forceCompState(client.UnitStateFailed, stopMsg)
```

Comment:
Unrelated to this PR, but more of an FYI. When the process restarts I don't think the agent clears any of the previous component state.

This causes the agent to send an initial CheckinExpected with non-zero config indexes and nil UnitExpectedConfigs since from the agent perspective nothing has changed, but to the new process it actually has no configuration yet and it's own config indexes have reset so it will try to apply the nil UnitExpectedConfigs.

For the Beats I believe this is the cause of the nil pointer dereferences that happen after an output change: https://github.com/elastic/beats/issues/34137#issuecomment-1367018103

Reply:
This additional panic and exit is going to count as a second restart in the 5s period, probably defeating the point of these changes on an intermittent basis.